Extensions/Firebase/firebasetools/D_authtools.js

```diff
@@ -0,0 +1,312 @@
+         */
+        changePassword(email, oldPassword, newPassword, callbackStateVariable) {
+            let credential = firebase.auth.EmailAuthProvider.credential(email, oldPassword);
+            gdjs.evtTools.firebase.auth.currentUser.reauthenticateWithCredential(credential)
+              .then(() => gdjs.evtTools.firebase.auth.currentUser.updatePassword(newPassword))
+              .then(() => callbackStateVariable.setString("ok"))
+              .catch(error => callbackStateVariable.setString(error.message));
+        },
+
+        /**
+         * Deletes the Current User.
+         * Use this when using basic auth.
+         * @param {string} email - Old Email for reauthentification.
+         * @param {string} password - Old password for reauthentification.
+         */
+        deleteUser(email, password) {
+            let credential = firebase.auth.EmailAuthProvider.credential(email, password);
+            gdjs.evtTools.firebase.auth.currentUser.reauthenticateWithCredential(credential)
+              .then(() => gdjs.evtTools.firebase.auth.currentUser.delete())
+              .then(() => callbackStateVariable.setString("ok"))
+              .catch(error => callbackStateVariable.setString(error.message));
+        },
+
+        /**
+         * Changes the user's email.
+         * Use this when using an external provider.
+         * @param {string} newEmail - New Email for the user.
+         * @param {boolean} sendVerificationEmail - Send a verification email to the old address before changing the email?
+         * @param {gdjs.Variable} callbackStateVariable - The variable where to store the result.
+         */
+        changeEmailProvider(newEmail, sendVerificationEmail, callbackStateVariable) {
+            let updater = sendVerificationEmail ? gdjs.evtTools.firebase.auth.currentUser.updateEmail : gdjs.evtTools.firebase.auth.currentUser.verifyBeforeUpdateEmail;
+
+            gdjs.evtTools.firebase.auth.currentUser.reauthenticateWithPopup(gdjs.evtTools.firebase.auth._currentProvider)
+              .then(() => updater(newEmail))
+              .then(() => callbackStateVariable.setString("ok"))
+              .catch(error => callbackStateVariable.setString(error.message));
+        },
+
+        /**
+         * Changes the user's password.
+         * Use this when using an external provider.
+         * @param {string} newPassword - New Password for the user.
+         * @param {gdjs.Variable} callbackStateVariable - The variable where to store the result.
+         */
+        changePasswordProvider(newPassword, callbackStateVariable) {
+            gdjs.evtTools.firebase.auth.currentUser.reauthenticateWithPopup(gdjs.evtTools.firebase.auth._currentProvider)
+              .then(() => gdjs.evtTools.firebase.auth.currentUser.updatePassword(newPassword))
+              .then(() => callbackStateVariable.setString("ok"))
+              .catch(error => callbackStateVariable.setString(error.message));
```

Comment:
In other extensions doing some kind of asynchronous operations (like the Facebook Instant Games API), the convention was to use a first variable to store the result and another variable to store the error, if any.

Not sure if it makes sense in your case - but think about it if you want to provide the same convention :)

Reply:
🤔 That would make lots of arguments and look maybe too intimidating?

Reply:
Maybe. Let's go with a single "state" variable for now then. It's ok here because what is updated is a global "authenticated" state.

If you need in another action to communicate both a result and a state as variables, then in this case do two variables (one containing the result, if any, and another one containing the state) (i.e: don't mix them).